### DISCUSSION TITLE
osu! may not run on certain Android devices.

### DISCUSSION THREAD

**Discussion by nphuracm** [ppy/osu#25373] [Q&A]
### Type

Crash to desktop

### Bug description

Device: Samsung Galaxy J3 Pro, China Telecom variant (SM-J3119)
 - SoC: Qualcomm Snapdragon 410
 - Arch: arm (armeabi-v7a) - despite having a ARMv8-capable CPU, this device runs mostly on ARMv7-based software.

The issue was first observed when I installed osu! on the device after installing a certain LineageOS 17.1 (Android 10) build. After the automatic screen rotation, the screen remained blank and no other screen activity were present. An ANR popup may appear at times.
It is possible that this issue is not operating system environment-dependent. I also attempted running osu! on the same device using the device manufacturer's factory image (Android 5.1.1), and the problem persisted. An ANR popup was also observed under said environment.

* I have only tested version 2023.1008.1 on the factory image. On LineageOS, the issue persisted across both 2023.1008.1 and 2023.1026.0.

### Screenshots or videos

Followed is an external recording of the directly observed behavior.
**I was not able to upload the 75MB video directly to GitHub. It is for this reason that the video is hosted [here](https://0x0.st/HtKE.MP4) instead. This link is not permanent.**

### Version

2023.1026.0

### Logs

Followed are the two log files which presence are _at least_ consistent across runs. There has been cases where an additional `network.log` and another log file were present, but they were unfortunately not preserved as of the time of this report. My milestone had varied across multiple runs.
[performance.log](https://github.com/ppy/osu/files/13256370/performance.log)
[runtime.log](https://github.com/ppy/osu/files/13256373/runtime.log)

**Comment by bdach** [comment]
The logs look like they were truncated suddenly, so we're not going to get much use out of them. Can you post logcat output?

**Comment by nphuracm** [comment]
[logcat.log](https://github.com/ppy/osu/files/13256467/logcat.log)
I did a `logcat -c` and redirected all subsequent messages that logcat would print to this file. I stopped recording a couple of minutes after launching the game.

**Comment by themcrgames100** [comment]
I have a Samsung Tablet sometimes it gets really laggy when having storyboards

**Comment by bdach** [comment]
I'm not sure what is going on there precisely but if I were to try and describe that it would be "broken beoyond belief".

> 11-04 19:25:02.574 15725 15725 W monodroid-assembly: Shared library 'EGL' not loaded, p/invoke 'eglGetCurrentContext' may fail

Nothing is going to be shown without egl present.

> 11-04 19:25:02.494 15725 15725 W monodroid-assembly: Shared library 'libc' not loaded, p/invoke 'uname' may fail

libc is not there either? What?

> 11-04 19:25:03.091 15725 15725 W monodroid-assembly: Shared library 'android' not loaded, p/invoke 'ANativeWindow_release' may fail

Okay what?

I cannot even think how to feasibly debug this but my bets are on this 'lineageOS' thing just not being supported period. Please post a logcat with the native image loaded if we're to continue investigating.

**Comment by nphuracm** [comment]
Current environment: Factory image; Samsung TouchWiz on Android 5.1.1.

[logcat2.log](https://github.com/ppy/osu/files/13258037/logcat2.log)
Still the same procedure. Ran `logcat -c` once, opened osu! from home screen and waited for a couple of minutes.
During this run, an additional `input.log` file was seen, though I don't believe it'd be of much use since it also looks truncated:
[input.log](https://github.com/ppy/osu/files/13258038/input.log)

UPDATE: ANR dialog observed.

**Comment by peppy** [comment]
That's a very old android version. While it still seems to be getting google play updates, security updates ended in 2018. I'm not sure we should be supporting this.

**Comment by bdach** [comment]
The root failure here is _probably_ this.

```
W/Adreno-EGL(  266): <qeglDrvAPI_eglSwapBuffers:3719>: EGL_BAD_SURFACE
E/SurfaceFlinger(  266): eglSwapBuffers(0x1, 0xb9064218) failed with 0x0000300d
```

But I don't know why it's failing like this.

> That's a very old android version. While it still seems to be getting google play updates, security updates ended in 2018. I'm not sure we should be supporting this.

My old samsung device from 2017 still runs the game. So I'm not sure what is going on there really.

Will probably convert this to a discussion until it becomes clear what (if anything) we can actually fix here.

**Comment by nphuracm** [comment]
Just chiming in for a bit, I do have a guess about the cause of the issue, that it could be something with the hardware. The SoC may be malfunctioning or possess some unknown PU/CPU quirks that prevent osu! from functioning. I'm suspecting this mostly because of the Adreno EGL error thrown in Android 5.1 and that the game client didn't seem to work even under multiple OS environments.

***Just for reference***(emphasis emphasis emphasis), I've also run things that sound like to depend on GL on that device:
 * [https://github.com/brunodev85/Winlator](Winlator) (wrapper for box86/64 with Wine): Crashes when attempting to start a new container.
 * [https://github.com/2394425147/maipaddx](AstroDX) (based on Unity): Works™, but significant rendering errors that prevent any usage for the app after changing a graphics quality setting.
